File: src.cs/alox/loggers/ConsoleLogger.cs
// #################################################################################################
//  cs.aworx.lox.loggers - ALox Logging Library
//
//  Copyright 2013-2018 A-Worx GmbH, Germany
//  Published under 'Boost Software License' (a free software license, see LICENSE.txt)
// #################################################################################################

using System;
using System.Runtime.CompilerServices;
using cs.aworx.lox.core.textlogger;
using cs.aworx.lib;
using cs.aworx.lib.strings;
using cs.aworx.lox.core;
using cs.aworx.lib.lang;


/** ************************************************************************************************
 * This is the C# namespace for the implementation of <em>logger classes</em> that are provided
 * by default with <b>%ALox Logging Library</b>.
 *
 * Developed by A-Worx GmbH and published under Boost Software License.
 **************************************************************************************************/
namespace cs.aworx.lox.loggers    {


/** ************************************************************************************************
 * A logger that logs all messages using the .Net class <em>System.Console</em>.
 * The name of the \e Logger defaults to "CONSOLE".
 *
 * \b %ALox text logger escape sequences (see class \ref cs.aworx.lox.ESC "ESC")
 * are removed and ignored.
 **************************************************************************************************/
public class ConsoleLogger : PlainTextLogger
{
    // #############################################################################################
    // Empty method stubs for release version (with no release logging)
    // #############################################################################################
    #if !(ALOX_DBG_LOG || ALOX_REL_LOG)
            public ConsoleLogger( String name= "CONSOLE" ){}
    #else
    // #############################################################################################
    // Fields
    // #############################################################################################

    /** ********************************************************************************************
     * Creates a ConsoleLogger.
     * @param name  (Optional) The name of the \e Logger, defaults to "CONSOLE"
     **********************************************************************************************/
    public ConsoleLogger( String name= null )
    : base( name, "CONSOLE", true )
    {
    }

    /** ********************************************************************************************
     * Start a new log line. Appends a new-line character sequence to previously logged lines.
     *
     * @param phase  Indicates the beginning or end of a log operation.
     * @return Always returns true.
     **********************************************************************************************/
    override
    protected bool notifyLogOp( Phase phase )
    {
        if ( phase == Phase.End )
        {
            #if !(ALOX_WP71 || ALOX_WP8)
                Console.WriteLine();
            #else
                Console.WriteLine();
            #endif
        }
        return true;
    }

    /** ********************************************************************************************
     * Write the given region of the given AString to the destination buffer.
     *
     * @param buffer   The string to write a portion of.
     * @param start    The start of the portion in \p{buffer} to write out.
     * @param length   The length of the portion in \p{buffer} to write out.
     * @return Always returns true.
     **********************************************************************************************/
    override
    protected bool logSubstring( AString buffer, int start, int length )
    {
        #if !(ALOX_WP71 || ALOX_WP8)
            Console.Write( buffer.Buffer(), start, length );
        #else
            Console.Write( buffer.ToString( 0, start, length );
        #endif
        return true;
    }

    /** ********************************************************************************************
     * Empty implementation, not needed for this class
     * @param phase  Indicates the beginning or end of a multi-line operation.
     **********************************************************************************************/
    override protected void notifyMultiLineOp( Phase phase )  {}

#endif // ALOX_DBG_LOG || ALOX_REL_LOG
} // class ConsoleLogger
} // namespace

File: src.java/alox/com/aworx/lox/loggers/ConsoleLogger.java
// #################################################################################################
//  com.aworx.lox.loggers - ALox Logging Library
//
//  Copyright 2013-2017 A-Worx GmbH, Germany
//  Published under 'Boost Software License' (a free software license, see LICENSE.txt)
// #################################################################################################

/** ************************************************************************************************
 *  This is the Java package for the implementation of <em>logger classes</em> that are provided
 *  by default with <b>%ALox Logging Library</b>.<p>
 *  Developed by A-Worx GmbH and published under Boost Software License.
 **************************************************************************************************/
package com.aworx.lox.loggers;

import java.io.PrintWriter;

import com.aworx.lib.config.Variable;
import com.aworx.lib.lang.Phase;
import com.aworx.lib.strings.AString;
import com.aworx.lox.ALox;
import com.aworx.lox.core.textlogger.PlainTextLogger;

/** ************************************************************************************************
 * A logger that logs all messages to the Java console using <em>System.out.print()</em> or
 * alternatively (if available) the <em>PrintWriter</em> object returned by <em>System.console()</em>.
 * The latter is used if it is available and will lead to a different character encoding
 * (typically on Windows operating system).
 * The configuration variable [ALOX_USE_SYSTEM_OUT_PRINT](../group__GrpALoxConfigVars.html)
 * can be used to force the use of <em>System.out.</em>.
 *
 * The name of the \e Logger defaults to "CONSOLE".
 **************************************************************************************************/
public class ConsoleLogger extends PlainTextLogger
{
    /**
     * Retrieved using <em>System.console().writer()</em>. See \ref ConsoleLogger "class description"
     * for more information.
     *
     * Can also be set from outside to redirect output to any <em>PrintWriter</em> type.
     */
    public       PrintWriter             writer;

    /** ********************************************************************************************
     * Creates a ConsoleLogger with the given name.
     *
     * @param name  The name of the \e Logger, defaults to "CONSOLE".
     **********************************************************************************************/
    public    ConsoleLogger( String    name )
    {
        super( name, "CONSOLE", true );
        constructor();
    }

    /** ********************************************************************************************
     * Creates a ConsoleLogger with the name "CONSOLE".
     **********************************************************************************************/
    public    ConsoleLogger( )
    {
        super( null, "CONSOLE", true );
        constructor();
    }

    /** ********************************************************************************************
     * Used by the constructors to perform common initialization tasks.
     **********************************************************************************************/
    protected  void  constructor()
    {
        // Enables logging to *System.console()* if such console is available. This might result in a
        // different character encoding (typically on Windows machines). */

        // checks a config variable to suppress using the system console instead of system.out
        Variable variable= new Variable( ALox.USE_SYSTEM_OUT_PRINT );
        ALox.config.load(variable);
        writer=  !variable.isTrue() && System.console() != null  ?  System.console().writer()
                                                                 :  null;
    }

    /** ********************************************************************************************
     * Start a new log line. Appends a new-line character sequence to previously logged lines.
     *
     * @param phase  Indicates the beginning or end of a log operation.
     * @return Always returns true.
     **********************************************************************************************/
    @Override
    protected boolean notifyLogOp( Phase phase )
    {
        if ( phase == Phase.END  )
        {
            if ( writer != null )
                writer.println();
            else
                System.out.println();
        }
        return true;
    }

    /** ********************************************************************************************
     * Write the given region of the given AString to the destination buffer.
     *
     * @param buffer   The string to write a portion of.
     * @param start    The start of the portion in \p buffer to write out.
     * @param length   The length of the portion in \p buffer to write out.
     * @return Always returns true.
     **********************************************************************************************/
    @Override
    protected boolean logSubstring( AString buffer, int start, int length )
    {
        char[]  buf=        buffer.buffer();
        if ( writer != null )
        {
            writer.write( buf, start, length );
        }
        else
        {
            int end= start + length;
            for (int i= start; i < end; i++ )
                System.out.print( buf[i] );
        }

        return true;
    }

    /** ********************************************************************************************
     * Empty, not needed here
     * @param phase  Indicates the beginning or end of a multi-line operation.
     **********************************************************************************************/
    @Override
    protected void notifyMultiLineOp (Phase phase)
    {
        /* nothing to do here */
    }

} // class ConsoleLogger
